config.go

```diff
@@ -0,0 +1,288 @@
+	if configs.ApkPath != "" {
+		log.Printf("- ApkPath: %s", configs.ApkPath)
+	}
+	if configs.AppPackageID != "" {
+		log.Printf("- AppPackageID: %s", configs.AppPackageID)
+	}
+	log.Printf("- TestTimeout: %f", configs.TestTimeout)
+	log.Printf("- FlakyTestAttempts: %d", configs.FlakyTestAttempts)
+	log.Printf("- DownloadTestResults: %t", configs.DownloadTestResults)
+	log.Printf("- DirectoriesToPull: %s", configs.DirectoriesToPullList)
+	log.Printf("- AutoGoogleLogin: %t", configs.AutoGoogleLogin)
+	log.Printf("- EnvironmentVariables: %s", configs.EnvironmentVariablesList)
+	log.Printf("- ObbFilesList: %s", configs.ObbFilesList)
+	log.Printf("- TestDevices:\n---")
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
+	if _, err := fmt.Fprintln(w, "Model\tAPI Level\tLocale\tOrientation\t"); err != nil {
+		failf("Failed to write in tabwriter, error: %s", err)
+	}
+	scanner := bufio.NewScanner(strings.NewReader(configs.TestDevicesList))
+	for scanner.Scan() {
+		device := scanner.Text()
+		device = strings.TrimSpace(device)
+		if device == "" {
+			continue
+		}
+
+		deviceParams := strings.Split(device, ",")
+
+		if len(deviceParams) != 4 {
+			continue
+		}
+
+		if _, err := fmt.Fprintln(w, fmt.Sprintf("%s\t%s\t%s\t%s\t", deviceParams[0], deviceParams[1], deviceParams[3], deviceParams[2])); err != nil {
+			failf("Failed to write in tabwriter, error: %s", err)
+		}
+	}
+	if err := w.Flush(); err != nil {
+		log.Errorf("Failed to flush writer, error: %s", err)
+	}
+
+	log.Printf("---")
+	log.Printf("- TestType: %s", configs.TestType)
+	// instruments
+	if configs.TestType == testTypeInstrumentation {
+		log.Printf("- TestApkPath: %s", configs.TestApkPath)
+		log.Printf("- InstTestPackageID: %s", configs.InstTestPackageID)
+		log.Printf("- InstTestRunnerClass: %s", configs.InstTestRunnerClass)
+		log.Printf("- InstTestTargets: %s", configs.InstTestTargets)
+		log.Printf("- UseOrchestrator: %t", configs.UseOrchestrator)
+	}
+
+	//robo
+	if configs.TestType == testTypeRobo {
+		log.Printf("- RoboInitialActivity: %s", configs.RoboInitialActivity)
+		log.Printf("- RoboScenarioFile: %s", configs.RoboScenarioFile)
+		log.Printf("- RoboDirectives: %s", configs.RoboDirectives)
+		log.Printf("- RoboMaxDepth: %s", configs.RoboMaxDepth)
+		log.Printf("- RoboMaxSteps: %s", configs.RoboMaxSteps)
+	}
+
+	// loop
+	if configs.TestType == "gameloop" {
+		log.Printf("- LoopScenarios: %s", configs.LoopScenarios)
+		log.Printf("- LoopScenarioLabels: %s", configs.LoopScenarioLabels)
+		log.Printf("- LoopScenarioNumbers: %s", configs.LoopScenarioNumbers)
+	}
+}
+
+func (configs *ConfigsModel) validate() error {
+	configs.migrate()
+
+	if strings.TrimSpace(configs.APIBaseURL) == "" {
+		if _, set := os.LookupEnv("BITRISE_IO"); !set {
+			log.Warnf("Warning: please make sure that Virtual Device Testing add-on is turned on under your app's settings tab.")
+		}
+		return fmt.Errorf("- APIBaseURL: required variable is not present")
```

Comment:
Please make it required for the struct as well.

Reply:
Did not add it, as that way the validation in  stepconf.Parse would fail, which is ok but the custom warning message will not be printed.